modules/ipp/scanner.go

```diff
@@ -0,0 +1,206 @@
+func getIPPURL(https bool, host string, port uint16, endpoint string) string {
+	var proto string
+	if https {
+		proto = "https"
+	} else {
+		proto = "http"
+	}
+	return proto + "://" + host + ":" + strconv.FormatUint(uint64(port), 10) + endpoint
+}
+
+func ippInContentType(resp http.Response) bool {
+	for _, t := range resp.Header["Content-Type"] {
+		if strings.Contains(t, CONTENT_TYPE) {
+			return true
+		}
+	}
+	return false
+}
+
+//TODO: Doesn't support TLS at all right now
+func (scanner *Scanner) grab(target zgrab2.ScanTarget) (int8, int8, *zgrab2.ScanError) {
+	//FIXME: This is not where this hostname assignment logic should live
+	host := target.Domain
+	if host == "" {
+		host = target.IP.String()
+	}
+	//FIXME: ?Should just use endpoint "/", since we get the same response as "/ipp" on CUPS??
+	uri := getIPPURL(scanner.config.IPPSecure, host, uint16(scanner.config.BaseFlags.Port), "/ipp")
+	b := getPrinterAttributesRequest(uri)
+	resp, err := http.Post(uri, CONTENT_TYPE, &b)
+	if err != nil {
+		//TODO: Create a descriptive error
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	if resp != nil && resp.Body != nil {
+		defer resp.Body.Close()
+	} else {
+		//TODO: Determine whether we need this error to avoid reading from Body
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	//FIXME: Maybe add something to handle redirects
+	//FIXME: Probably return the whole response for further inspection by ztag, rather
+	//         than grabbing first 2 bytes. In that case, implement maxRead like http module
+
+	//Check to make sure that the repsonse received is actually IPP
+	//Content-Type header matches is sufficient
+	//HTTP on port 631 is sufficient
+	//Still record data in the case of protocol error to see what that data looks like
+
+	//TODO: Record server-header version numbers
+	//protocols := resp.Header["Server"])
+	var version int16
+	if err := binary.Read(resp.Body, binary.BigEndian, &version); err != nil {
+		return 0, 0, zgrab2.NewScanError(zgrab2.SCAN_UNKNOWN_ERROR, err)
+	}
+	return int8(version >> 8), int8(version & 0xff), nil
+}
+
+// Scan TODO: describe how scan operates
+//1. Send a request (currently get-printer-attributes)
+//2. Take in that response & read out version numbers
+func (scanner *Scanner) Scan(target zgrab2.ScanTarget) (zgrab2.ScanStatus, interface{}, error) {
+	// TODO: use Connection again, at least when implementing TLS
+	major, minor, err := scanner.grab(target)
+	//FIXME: Triggering even though error IS nil
+	//FIXME: This is a sloppy bodge to handle the issue
+	if major == 0 && minor == 0 && err != nil {
+		//TODO: Consider mimicking HTTP Scan's retryHTTPS functionality
+		//TODO: Create more detailed error message?
+		return zgrab2.TryGetScanStatus(err), nil, err
+	}
+	results := &ScanResults{}
+	results.MajorVersion = major
+	results.MinorVersion = minor
+	return zgrab2.SCAN_SUCCESS, results, nil
```

Comment:
It looks like you're throwing away `err` if it was able to detect a major/minor; it looks like the intent was something like `return zgrab2.TryGetScanStatus(err), results, err`; if `err == nil`, that will be identical to what you have now, but if there is an error, it will still return the partial results, alongside the error and a status.